.github/workflows/build-wheels.yaml

```diff
@@ -3,83 +3,291 @@ run-name: build-wheels
+
+      - if: ${{ steps.restore_linux_wheels.outputs.cache-hit != 'true' }}
+        name: Test wheels across Python 3.9–3.12
         run: |
           set -euxo pipefail
 
           for PYVER in 3.9 3.10 3.11 3.12; do
-            TAG="cp${PYVER/./}"
-
-            echo "Testing on Python $PYVER (${{ matrix.arch }}) with tag $TAG"
+            PYVER_SHORT=${PYVER/./}
+            echo "Testing on Python $PYVER"
 
             docker run --rm \
-              -v $(pwd)/${{ matrix.wheel_dir }}:/wheels \
+              -v $(pwd)/wheelhouse:/wheels \
               -v $(pwd)/tests:/tests \
               python:$PYVER-slim /bin/bash -c "
                 set -euxo pipefail
-                apt-get update && apt-get install -y gcc python3-dev
-                python -m pip install --upgrade pip
-                WHEEL=\$(ls /wheels/*${TAG}*.whl | head -n 1)
+                WHEEL=\$(ls /wheels/*cp${PYVER_SHORT}*manylinux*.whl)
                 echo 'Installing wheel: ' \$WHEEL
                 python -m pip install \$WHEEL
                 python -c 'import genome_kit; print(genome_kit.__version__)'
                 ls /tests/test_*.py | sed 's/\.py$//' | sed 's/\/tests\//tests./' | CI=1 xargs --verbose -I {} python -m unittest {}
               "
           done
+
+      - if: ${{ steps.restore_linux_wheels.outputs.cache-hit != 'true' }}
+        name: cache linux wheels
+        uses: actions/cache/save@v4
+        with:
+          # hashFiles() would produce a different key due to setup/*.pyc files created
+          key: ${{ steps.restore_linux_wheels.outputs.cache-primary-key }}
+          path: wheelhouse/*.whl
+
+
+  build-wheel-macos:
+    name: Build macos wheels
+    runs-on: macos-14
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: try to restore the macos wheels cache
+        id: restore_macos_wheels
+        uses: actions/cache/restore@v4
+        with:
+          key: macos-wheels-${{ hashFiles('.github/workflows/build-wheels.yaml', '.github/actions/**', 'src/**', 'genome_kit/**', '_pyproject.toml', 'setup.py', 'setup/**') }}
+          path: wheelhouse/*.whl
+
+      # skip remaining steps on cache hit
+
+      - if: ${{ steps.restore_macos_wheels.outputs.cache-hit != 'true' }}
+        name: Rename pyproject.toml
+        run: mv _pyproject.toml pyproject.toml
+
+      - if: ${{ steps.restore_macos_wheels.outputs.cache-hit != 'true' }}
+        name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.12'
+
+      - if: ${{ steps.restore_macos_wheels.outputs.cache-hit != 'true' }}
+        name: Install cibuildwheel
+        run: python -m pip install cibuildwheel==2.23.3
+
+      - if: ${{ steps.restore_macos_wheels.outputs.cache-hit != 'true' }}
+        name: Build wheels
+        run: python -m cibuildwheel --output-dir wheelhouse && ls -l wheelhouse/
+        env:
+          GK_BUILD_WHEELS: "1"
+          CIBW_ARCHS_MACOS: x86_64 arm64
```

Comment:
is this a universal build?  or it does both, if so, there's no key on the architecture?

if this is all justified, would be nice to have a comment on why it's different than arm/x86 linux